### PR TITLE
Create a new job for syncronizing secrets from multiple vault engines to k8s

### PR DESCRIPTION
**Pull request by phanyzewski**
new command which defines a default and an overriding vault path for synchronizing secrets from multiple vault engines

### CODE REVIEW

File: cmd/vault-staging-k8s-secret/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+RUN go version \
+ && go mod download \
+ && go mod verify
+
+COPY . .
+
+RUN env GOOS=linux GOARCH=amd64 go build -o vault-staging-k8s-secret
+
+CMD ./entrypoint.sh
```
Comment (ttacon) [inline cmd/vault-staging-k8s-secret/Dockerfile:28]: This doesn't block this PR, but do we want to add a secondary build step that is a slim OS without code and dependencies (that we copy the binary over to)?

File: cmd/vault-staging-k8s-secret/main.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"fmt"
+	"os"
+
+	"github.com/teamsnap/vault-key/pkg/k8s"
+	"github.com/teamsnap/vault-key/pkg/vault"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// var env = map[string]map[string]string{}
```
Comment (ttacon) [inline cmd/vault-staging-k8s-secret/main.go:14]: Minor item, let's remove this line.

```suggestion
```

File: cmd/vault-staging-k8s-secret/main.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		EncoderConfig:    zap.NewDevelopmentEncoderConfig(),
+		OutputPaths:      []string{"stderr"},
+		ErrorOutputPaths: []string{"stderr"},
+	}
+
+	logger, err := cfg.Build()
+	if err != nil {
+		panic(err)
+	}
+	cfg.Level.SetLevel(zap.ErrorLevel)
```
Comment (ttacon) [inline cmd/vault-staging-k8s-secret/main.go:29]: Do we want this to be configurable?

File: cmd/vault-staging-k8s-secret/main.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	run(context.Background(), logger)
+}
+
+func run(ctx context.Context, lgr *zap.Logger) {
+	defaultEngine := os.Getenv("VAULT_SECRET")
```
Comment (ttacon) [inline cmd/vault-staging-k8s-secret/main.go:35]: From a testability standpoint, how do you feel about passing in this (and the other envvars) via a config object?

File: cmd/vault-staging-k8s-secret/main.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	secrets := map[string]map[string]string{}
+	vault.GetSecrets(ctx, &secrets, verifiedEngines)
+
+	lgr.Info("creating overrides", zap.String("overrides", fmt.Sprintf("%v", secrets[override])))
+	mergedSecrets := secrets[override]
+
+	lgr.Info("setting defaults")
+	for k, v := range secrets[defaultEngine] {
+		if _, ok := mergedSecrets[k]; !ok {
+			mergedSecrets[k] = v
```
Comment (ttacon) [inline cmd/vault-staging-k8s-secret/main.go:76]: We should add a note that the first secret wins out, or an affordance for reporting when conflicting secrets are detected and are being dropped.

Comment (phanyzewski) [reply cmd/vault-staging-k8s-secret/main.go:76]: On L#70 I was dumping the entire override secret, but I like logging each one individually here better 👍 

Makes the output a little bit noisier but I think the verbosity when troubleshooting would be useful.

File: cmd/vault-staging-k8s-secret/Dockerfile (hunk 1 above)
Comment (phanyzewski) [reply cmd/vault-staging-k8s-secret/Dockerfile:29]: Took a few iterations but I think I have a decent looking dockerfile now.  

before / after 
```sh
01:58:20 PM ❯ docker image ls
REPOSITORY                                             TAG                 IMAGE ID       CREATED          SIZE
us.gcr.io/staging-205121/vault-key/staging-k8-secret   0.0.1-fat           8548c7cac49d   13 minutes ago   4.14GB
us.gcr.io/staging-205121/vault-key/staging-k8-secret   0.0.1               9762d4b4f75a   35 minutes ago   769MB

Comment (phanyzewski) [reply cmd/vault-staging-k8s-secret/Dockerfile:29]: after pushing to gcr it becomes a much slimmer file ~197 MB

File: cmd/vault-staging-k8s-secret/config.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	override      string
+	k8sNamespace  string
+	lgr           *zap.Logger
+}
+
+func newConfig() *config {
+	lgr := newLogger()
+
+	defaultEngine := os.Getenv("VAULT_SECRET")
+	lgr.Debug("vault-path to the default engine", zap.String("VAULT_SECRET", defaultEngine))
```
[the remark this answers is not in the document]
Comment (phanyzewski) [reply cmd/vault-staging-k8s-secret/config.go]: I also reorganized this to read in the entire required config before making a call to fatal

File: cmd/vault-staging-k8s-secret/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
+
+set -e
+
```
Comment (stevenaldinger) [inline cmd/vault-staging-k8s-secret/entrypoint.sh:4]: could possibly do something like:
```
if [ "${DEBUG}" = "true" ]
then
  set -x
fi
```

**Pull request merged**